packages/validator/src/services/prepareBeaconProposer.ts

```diff
@@ -51,4 +75,64 @@ export class PrepareBeaconProposerService {
       ),
     }));
   }
+
+  private registerValidator = async (epoch: Epoch): Promise<void> => {
+    if (epoch < this.config.BELLATRIX_FORK_EPOCH - 1) return;
+    this.getValidatorRegistrations(epoch, this.indicesService.getAllLocalIndices())
+      .then((registrations) => this.registerValidatorInChunks(registrations))
+      .catch((e: Error) => {
+        this.logger.error("Error on registerValidator", {epoch}, e);
+      });
+
+    // At the same time fetch any remaining unknown validator indices, then poll duties for those newIndices only
+    this.indicesService
+      .pollValidatorIndices()
+      .then((newIndices) =>
+        this.getValidatorRegistrations(epoch, newIndices).then((registrations) =>
+          this.registerValidatorInChunks(registrations)
+        )
+      )
+      .catch((e: Error) => {
+        this.logger.error("Error on poll indices and prepareBeaconProposer", {epoch}, e);
+      });
+  };
+
+  private registerValidatorInChunks = async (
+    registrations: bellatrix.SignedValidatorRegistrationV1[]
+  ): Promise<void> => {
+    const registrationChunks = spliceIntoChunks(registrations, REGISTRATION_CHUNK_SIZE);
+    await Promise.all(
+      registrationChunks.map((registrationChunk) => this.api.validator.registerValidator(registrationChunk))
+    );
+    return;
+  };
+
+  private getValidatorRegistrations = async (
+    epoch: Epoch,
+    indices: number[]
+  ): Promise<bellatrix.SignedValidatorRegistrationV1[]> => {
+    const slot = epoch * SLOTS_PER_EPOCH;
+    return Promise.all(
+      indices.map((validatorIndex) => {
+        const pubkeyHex = this.indicesService.index2pubkey.get(validatorIndex);
+        if (!pubkeyHex) throw Error(`Pubkey lookup failure for validatorIndex=${validatorIndex}`);
+        const feeRecipient = this.validatorStore.feeRecipientByValidatorPubkey.getOrDefault(pubkeyHex);
+        return this.validatorStore.signValidatorRegistration(
+          fromHexString(pubkeyHex),
+          fromHexString(feeRecipient),
+          this.validatorStore.gasLimit,
+          slot
+        );
+      })
+    );
+  };
+}
+
+function spliceIntoChunks<T>(arr: T[], chunkSize: number): T[][] {
```

Comment:
Move to utils and add many unit tests

Reply:
replaced with already defined (with tests) `batchItems`